Add `bytes` type validator

dswij opened the pull request:
Part of #9

Add `bytes` type validator

samuelcolvin commented on pydantic_core/_types.py:
```diff
@@ -163,6 +171,7 @@ class UnionSchema(TypedDict):
     BareType,
     AnySchema,
     BoolSchema,
+    BytesSchema,
```
Please add a case to `tests/test_typing.py`.

samuelcolvin commented on src/errors/kinds.rs:
```diff
@@ -107,6 +107,14 @@ pub enum ErrorKind {
+    #[strum(message = "Value must be a valid bytes")]
+    BytesType,
+    #[strum(message = "Bytes must have at least {min_length} characters")]
+    BytesTooShort,
```
You might want to choose a generic word like "Input" instead of "Bytes/Stirng"

dswij replied:
> as per #76, I think we should have one `TooShort` error which we us for strings and bytes.
> 
> You can either leave this and we'll fix them all in one PR, or do it here.

Let's fix them all in one PR. I can get to it after this one.

samuelcolvin commented on tests/validators/test_bytes.py:
```diff
@@ -0,0 +1,27 @@
+from pydantic_core import SchemaValidator, ValidationError
+
+from ..conftest import Err
+
```
you probably want to start with a general parametrised test which shows which types work and which do not.

You will also need to add a test (parametrised or not) which uses `py_or_json`) and therefore tests bytes fields from JSON.

samuelcolvin reviewed (commented):
This is looking amazing, thank you so much.

Let me know if you have any questions.

dswij commented on src/input/input_json.rs:
```diff
@@ -114,6 +114,15 @@ impl Input for JsonInput {
+    fn strict_bytes(&self) -> ValResult<Vec<u8>> {
+        match self {
+            JsonInput::String(s) => Ok(s.clone().into_bytes()),
+            JsonInput::Int(int) => Ok(int.to_ne_bytes().to_vec()),
```
I stumbled on this issue too while writing this. After giving this a thought, there's really no way to know exactly what value the user wants. I'd imagine arbitrarily choosing coercion will add frustration.

> if there's a (single) obvious representation in the field type AND converting to that looses no information, coerce, otherwise error.

This sounds like a good rule of thumb. If there's no single obvious representation, letting the user work around it will remove ambiguity.

samuelcolvin replied:
Agreed 👍.

dswij commented on src/input/input_abstract.rs:
```diff
@@ -44,4 +44,10 @@ pub trait Input: fmt::Debug + ToPy + ToLocItem {
         self.strict_set()
     }
+
+    fn strict_bytes(&self) -> ValResult<Vec<u8>>;
```
I don't think we can use a slice here. We're creating a new value when we parse the input, and we need an owner for the new value. 

Creating a slice here will be temporary and then dropped immediately.

samuelcolvin commented:
> The benchmark doesn't look too good if I understand it correctly:

looks like something  must be wrong, I'll have a look.

dswij commented:
> I've added the return_enums.rs file as we'll want to do this for a number of other types in future.

> I've also cleaned up your benchmarks somewhat, you had a benchmark with just a list with a single element which wasn't doing anything.

Sounds great. Is there anything else missing for the bytes validator?

samuelcolvin commented:
thanks so much for this.
